internal/testhelpers/session.go

```diff
@@ -136,92 +178,261 @@ func NewHTTPClientWithSessionToken(t *testing.T, reg *driver.RegistryDefault, se
 	}
 }
 
-func NewHTTPClientWithArbitrarySessionToken(t *testing.T, reg *driver.RegistryDefault) *http.Client {
-	req := x.NewTestHTTPRequest(t, "GET", "/sessions/whoami", nil)
-	s, err := session.NewActiveSession(req,
-		&identity.Identity{ID: x.NewUUID(), State: identity.StateActive},
-		NewSessionLifespanProvider(time.Hour),
-		time.Now(),
-		identity.CredentialsTypePassword,
-		identity.AuthenticatorAssuranceLevel1,
-	)
-	require.NoError(t, err, "Could not initialize session from identity.")
+func EnsureAAL(t *testing.T, c *http.Client, ts *httptest.Server, aal string, methods ...string) {
+	res, err := c.Get(ts.URL + session.RouteWhoami)
+	require.NoError(t, err)
+	sess := x.MustReadAll(res.Body)
+	require.NoError(t, res.Body.Close())
+	assert.EqualValues(t, aal, gjson.GetBytes(sess, "authenticator_assurance_level").String())
+	for _, method := range methods {
+		assert.EqualValues(t, method, gjson.GetBytes(sess, "authentication_methods.#(method=="+method+").method").String())
+	}
+	assert.Len(t, gjson.GetBytes(sess, "authentication_methods").Array(), 1+len(methods))
+}
 
-	return NewHTTPClientWithSessionToken(t, reg, s)
+func NewAuthorizedTransport(t *testing.T, reg *driver.RegistryDefault, sess *session.Session) *x.TransportWithHeader {
+	maybePersistSession(t, reg, sess)
+
+	return x.NewTransportWithHeader(http.Header{
+		"Authorization": {"Bearer " + sess.Token},
+	})
 }
 
-func NewHTTPClientWithArbitrarySessionCookie(t *testing.T, reg *driver.RegistryDefault) *http.Client {
-	req := x.NewTestHTTPRequest(t, "GET", "/sessions/whoami", nil)
-	s, err := session.NewActiveSession(req,
-		&identity.Identity{ID: x.NewUUID(), State: identity.StateActive},
-		NewSessionLifespanProvider(time.Hour),
-		time.Now(),
-		identity.CredentialsTypePassword,
-		identity.AuthenticatorAssuranceLevel1,
+// A builder struct to help customize an authenticated http client
+type HTTPClientBuilder struct {
+	t        *testing.T
+	request  *http.Request
+	identity *identity.Identity
+	session  *session.Session
+}
+
+func NewHTTPClientBuilder(t *testing.T) *HTTPClientBuilder {
+	return &HTTPClientBuilder{t: t}
+}
+
+// Raise a test error if the request hasn't been assigned yet.
+func (c *HTTPClientBuilder) nilRequestCheck() {
+	if c.request == nil {
+		c.t.Errorf("creating a Session requires a request to be present on the builder, but request was nil")
+	}
+}
+
+// Raise a test error if the identity hasn't been assigned yet.
+func (c *HTTPClientBuilder) nilIdentityCheck() {
+	if c.request == nil {
+		c.t.Errorf("creating a Session requires an identity to be present on the builder, but identity was nil")
+	}
+}
+
+// Raise a test error if the session hasn't been assigned yet.
+func (c *HTTPClientBuilder) nilSessionCheck() {
+	if c.session == nil {
+		c.t.Errorf("consuming the HTTPClientBuilder requires a session to be present on the builder, but session was nil")
+	}
+}
```

Comment:
Nil checks all error out immediately because it means the test has been written incorrectly.